tensorflow/core/kernels/matmul_op_impl.h

```diff
@@ -359,115 +401,328 @@ struct LaunchBatchMatMul<GPUDevice, Scalar> {
+        }
+        for (int64_t i = 0; i < batch_size; ++i) {
+          c_device_memory.push_back(AsDeviceMemory(c_base_ptr + i * m * n));
+          a_ptrs.push_back(&a_device_memory[a_batch_indices[i]]);
+          b_ptrs.push_back(&b_device_memory[b_batch_indices[i]]);
+          c_ptrs.push_back(&c_device_memory.back());
+        }
 
-    // Blas does
-    // C = A x B
-    // where A, B and C are assumed to be in column major.
-    // We want the output to be in row-major, so we can compute
-    // C' = B' x A', where ' stands for transpose (not adjoint).
-    // TODO(yangzihao): Choose the best of the three strategies using autotune.
-    if (batch_size == 1) {
-      // This is a regular matrix*matrix or matrix*vector multiply. Avoid the
-      // overhead of the scratch allocator and the batch interface.
-      if (n == 1 &&
-          blas_transpose_b != se::blas::Transpose::kConjugateTranspose &&
-          blas_transpose_a != se::blas::Transpose::kConjugateTranspose) {
-        // This is a matrix*vector multiply so use GEMV to compute A * b.
-        // Here we are multiplying in the natural order, so we have to flip
-        // the transposition flag to compensate for the tensor being stored
-        // row-major. Since GEMV doesn't provide a way to just conjugate an
-        // argument, we have to defer those cases to GEMM below.
-        auto gemv_trans_a = blas_transpose_a == se::blas::Transpose::kTranspose
-                                ? se::blas::Transpose::kNoTranspose
-                                : se::blas::Transpose::kTranspose;
+        BlasScratchAllocator scratch_allocator(context, max_scratch_size);
         bool blas_launch_status =
             stream
-                ->ThenBlasGemv(gemv_trans_a, adj_x || trans_x ? m : k,
-                               adj_x || trans_x ? k : m,
-                               static_cast<Coefficient>(1.0), *(a_ptrs[0]),
-                               adj_x || trans_x ? m : k, *(b_ptrs[0]), 1,
-                               static_cast<Coefficient>(0.0), c_ptrs[0], 1)
+                ->ThenBlasGemmBatchedWithScratch(
+                    blas_transpose_b, blas_transpose_a, n, m, k,
+                    static_cast<Coefficient>(1.0), b_ptrs,
+                    adj_y || trans_y ? k : n, a_ptrs, adj_x || trans_x ? m : k,
+                    static_cast<Coefficient>(0.0), c_ptrs, n, batch_size,
+                    &scratch_allocator)
                 .ok();
         if (!blas_launch_status) {
           context->SetStatus(errors::Internal(
-              "Blas xGEMV launch failed : a.shape=", in_x.shape().DebugString(),
+              "Blas xGEMMBatched launch failed: a.shape=",
+              in_x.shape().DebugString(),
               ", b.shape=", in_y.shape().DebugString(), ", m=", m, ", n=", n,
-              ", k=", k));
+              ", k=", k, ", batch_size=", batch_size));
+        }
+      }
+      return;
```

Comment:
Probably no point to return since once we fall out of this block, the function ends anyways. (Same for the same spot in the following function)